server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -107,6 +101,28 @@ public TransportFieldCapabilitiesAction(
         this.ccsCheckCompatibility = SearchService.CCS_VERSION_CHECK_SETTING.get(clusterService.getSettings());
     }
 
+    @Override
+    public void registerWithTransport(TransportActionRegister transportAction) {
```

Comment:
I think these register methods shouldn't be on the instance - it only really makes sense to call them when constructing the action in the first place. Instead could we try having a static method (within each action class) which first creates the action instance and then registers it with the appropriate registries.

Reply:
You are right, I also thought this was way more suitable as a static method.
I considered it, and chose to try with instance methods first because it seemed like a pattern more difficult to enforce (there are no static "virtual" interface methods in Java - like C# 11- right?).
But maybe we do not need to really "enforce" it?
I will give it a go and see how it looks like!

Reply:
I thought a bit more about the factory alternative:
```
interface ActionFactory {
   TransportAction createAndRegister(Registry....);
}
class SomeAction extends TransportAction {
     class Factory implements ActionFactory {
        public SomeActionFactory(/*ctor args: services etc*/) {
        }
        @Override
        public TransportAction createAndRegister(Registry registry) {
            var action = new SomeAction(/*pass args*/);
            registry.registerWithNodeClient(executor, action);
            ....
            return action;
        }
     }
}
```
on the registration side:
```
actionRegistry.register(new SomeAction.Factory(/*all args*/);
```
`register` will call the factory passing the appropriate lambdas/Registry implementations for TransportService and NodeClient.
It's a compromise; it looks clean but if the action uses some services there is a lot of pass-and-store (which is pretty automatic code - maybe we can even use a record). What do you think?

Reply:
We do not even need to return the TransportAction really, just pass it down to the Registry

Reply:
I'm not sure I understand the benefit of creating an instance of a factory in order to pass it to the registry for the registry to invoke its build method, over a plain old static method invoked as `SomeAction.createAndRegister(actionRegistry)`.

Reply:
or maybe `SomeAction.createAndRegister(bunchOfServices, actionRegistry)`, or maybe the action registry is itself a service in the `bunchOfServices`, various options here.

Reply:
It does not make a great difference, I agree. I felt it was a bit cleaner to separate what is needed to create the action object from what is needed to register it. 
My intention was to try and enforce that the registration actually happens (it's not "forgotten") by doing it in the registry and provide some guidance to action developers on what is needed to create an action (in the form of the factory interface). 
But it is a small thing, since we cannot enforce the fact that creation needs to go through the factory anyway.

Of those you propose, I would with `SomeAction.createAndRegister(bunchOfServices, actionRegistry)`; those will be some long signatures (actionRegistry could be "lost" (visually) at the end of it, another thing I did not like too much, but I know this is just a personal preference), but I feel it's better to have to write one thing. If I need to type both always together, keeping creation and registration separate looks more brittle.